homeassistant/components/nmbs/sensor.py

```diff
@@ -162,16 +164,17 @@ def update(self) -> None:
         """Set the state equal to the next departure."""
         liveboard = self._api_client.get_liveboard(self._station)
 
-        if (
-            liveboard is None
-            or liveboard.get("departures") is None
-            or liveboard.get("departures").get("number") is None
-            or liveboard.get("departures").get("number") == "0"
-            or liveboard.get("departures").get("departure") is None
-        ):
+        if liveboard == API_FAILURE:
+            _LOGGER.warning("API fails in NMBSLiveBoard")
             return
 
-        next_departure = liveboard["departures"]["departure"][0]
+        try:
+            if liveboard.get("departures").get("number") == "0":
+                return
+            next_departure = liveboard["departures"]["departure"][0]
+        except Exception as exc:  # pylint: disable=broad-except
```

Comment:
You shouldn't use a broad exception here.
There are much better ways to handle this.

I said before that you should post some sample logs for why this is necessary...

I suggested something closer to:
```python
    liveboard = self._api_client.get_liveboard(self._station)

    if liveboard == API_FAILURE:
        _LOGGER.warning("API failed in NMBSLiveBoard")
        return

    if liveboard.get("departures") is None or ...:
        _LOGGER.warning("API returned invalid data: %r", liveboard)
        return

    _LOGGER.debug("API returned data: %r", liveboard)
    next_departure = liveboard["departures"]["departure"][0]
```

Reply:
Here is a sample of today's log:
```
2023-04-05 14:59:23.993 WARNING (SyncWorker_3) [custom_components.nmbs.sensor] API fails in NMBSSensor
2023-04-05 15:02:24.045 WARNING (SyncWorker_3) [custom_components.nmbs.sensor] API fails in NMBSSensor
2023-04-05 15:04:35.075 WARNING (SyncWorker_6) [custom_components.nmbs.sensor] API fails in NMBSLiveBoard
2023-04-05 15:08:24.453 WARNING (SyncWorker_8) [custom_components.nmbs.sensor] API fails in NMBSSensor
2023-04-05 18:56:18.793 DEBUG (SyncWorker_8) [custom_components.nmbs.sensor] KeyError('connection') in NMBSSensor | API return data: {'error': 500, 'message': 'Could not get data: This request failed due to internal errors.. Please report this issue at https://github.com/irail/irail/issues/new'}
2023-04-05 18:56:43.763 DEBUG (SyncWorker_6) [custom_components.nmbs.sensor] KeyError('connection') in NMBSSensor | API return data: {'error': 500, 'message': 'Could not get data: This request failed due to internal errors.. Please report this issue at https://github.com/irail/irail/issues/new'}
```

Reply:
I wonder if pyRail should not be rewritten.

Reply:
...and this morning:
```
2023-04-06 03:03:22.058 DEBUG (SyncWorker_8) [custom_components.nmbs.sensor] KeyError('connection') in NMBSSensor | API return data: {'error': 500, 'message': 'Could not get data: This request failed due to internal errors.. Please report this issue at https://github.com/irail/irail/issues/new'}
2023-04-06 03:03:46.999 DEBUG (SyncWorker_1) [custom_components.nmbs.sensor] KeyError('connection') in NMBSSensor | API return data: {'error': 500, 'message': 'Could not get data: This request failed due to internal errors.. Please report this issue at https://github.com/irail/irail/issues/new'}
2023-04-06 05:34:29.997 DEBUG (SyncWorker_7) [custom_components.nmbs.sensor] KeyError('connection') in NMBSSensor | API return data: {'error': 500, 'message': 'Could not get data: This request failed due to internal errors.. Please report this issue at https://github.com/irail/irail/issues/new'}
2023-04-06 05:34:52.869 DEBUG (SyncWorker_0) [custom_components.nmbs.sensor] KeyError('connection') in NMBSSensor | API return data: {'error': 500, 'message': 'Could not get data: This request failed due to internal errors.. Please report this issue at https://github.com/irail/irail/issues/new'}
2023-04-06 05:37:53.157 WARNING (SyncWorker_7) [custom_components.nmbs.sensor] API fails in NMBSSensor
2023-04-06 05:37:58.219 WARNING (SyncWorker_8) [custom_components.nmbs.sensor] API fails in NMBSLiveBoard
2023-04-06 05:48:31.015 WARNING (SyncWorker_0) [custom_components.nmbs.sensor] API fails in NMBSLiveBoard
2023-04-06 05:57:28.194 WARNING (SyncWorker_8) [custom_components.nmbs.sensor] API fails in NMBSLiveBoard
2023-04-06 06:01:29.069 WARNING (SyncWorker_1) [custom_components.nmbs.sensor] API fails in NMBSLiveBoard
```

Reply:
There are three scenarios:
- `if liveboard == API_FAILURE` / `if connections == API_FAILURE` => that is a libray error => log warning, and exit
- `if not liveboard.get("departures")` / `if not connections.get("connection")`  => that is a libray error => log warning with the data, and exit
- remaining scenarios need to be handled in Home Assistant, and you should not use a broad exception.

Here is a new code sample:
```python
    liveboard = self._api_client.get_liveboard(self._station)

    if liveboard == API_FAILURE:
        _LOGGER.warning("API failed in NMBSLiveBoard")
        return

    if not (departures := liveboard.get("departures")):
        _LOGGER.warning("API returned invalid departures: %r", liveboard)
        return

    _LOGGER.debug("API returned departures: %r", departures)
    next_departure = departures["departure"][0]
```